source/event_loop.c

```diff
@@ -114,35 +158,87 @@ struct aws_event_loop_group *aws_event_loop_group_new(
 
     aws_global_thread_creator_increment();
 
+    if (pin_threads) {
+        aws_mem_release(alloc, usable_cpus);
+    }
+
     return el_group;
 
 on_error:
 
+    aws_mem_release(alloc, usable_cpus);
     s_aws_event_loop_group_shutdown_sync(el_group);
     s_event_loop_group_thread_exit(el_group);
 
     return NULL;
 }
 
-static struct aws_event_loop *default_new_event_loop(
-    struct aws_allocator *allocator,
+struct aws_event_loop_group *aws_event_loop_group_new(
+    struct aws_allocator *alloc,
     aws_io_clock_fn *clock,
+    uint16_t el_count,
+    aws_new_event_loop_fn *new_loop_fn,
+    void *new_loop_user_data,
+    const struct aws_shutdown_callback_options *shutdown_options) {
+
+    AWS_ASSERT(new_loop_fn);
+    AWS_ASSERT(el_count);
+
+    return s_event_loop_group_new(alloc, clock, el_count, 0, false, new_loop_fn, new_loop_user_data, shutdown_options);
+}
+
+static struct aws_event_loop *s_default_new_event_loop(
+    struct aws_allocator *allocator,
+    const struct aws_event_loop_options *options,
     void *user_data) {
 
     (void)user_data;
-    return aws_event_loop_new_default(allocator, clock);
+    return aws_event_loop_new_default_with_options(allocator, options);
 }
 
 struct aws_event_loop_group *aws_event_loop_group_new_default(
     struct aws_allocator *alloc,
     uint16_t max_threads,
     const struct aws_shutdown_callback_options *shutdown_options) {
     if (!max_threads) {
-        max_threads = (uint16_t)aws_system_info_processor_count();
+        uint16_t processor_count = (uint16_t)aws_system_info_processor_count();
+        /* cut them in half to avoid using hyper threads for the IO work. */
+        max_threads = processor_count > 1 ? processor_count / 2 : processor_count;
```

Comment:
This assumes that every type of CPU has hyperthreads
Graviton2 does not have hyperthreads

Reply:
I don't know how else you'd detect it. You have any ideas on it?

Reply:
On windows, it's apparently not too hard? [GetLogicalProcessorInformation](https://docs.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-getlogicalprocessorinformation)

Not finding an easy way for posix: [stackoverflow discussion](https://stackoverflow.com/questions/5118110/how-to-get-number-of-physical-cpus-in-linux-using-c)